### PR TITLE
Proposal: unban 'thiserror'

### PR DESCRIPTION
**Pull request by guswynn**
There is a reason why we have so many exceptions for the use of `thiserror` in our `deny.toml`. Here are some reasons why we should unban it:

- It is well-loved, well-tested, and there are plenty of examples of how to use it DSL in its docs and in the greater rust community
- It reduces boilerplate, and localizes information about errors. (see `CreateTopicError` in the attached code) The `Display` impl, the `From` impls, and the `Error` impls are all in 1 place.
- It is flexible enough for complex errors (See the `ErrorKind` wrapper for the coord catalog `Error` in the attached pr, and the display string for the `AmbiguousRename` variant)
- It reduces the prevalence of subtle bugs, like missing a correct `source` impl (see `S3Error` in the attached pr)

Additionally:
- It will make using `backtrace`'s trivial when they are stabilized (just `#[backtrace]`)
- Reduces the strain and boilerplate of writing a custom error, and will reduce our use of just `String` for `Err`'s in the codebase, producing more structured code.
- Works well with our use of `anyhow`. `thiserror` makes it easy to just add a:
```
#[error(transparent)]
Other(#[from] anyhow::Error)
```
to error enums, making `?` work very nicely!

### CODE REVIEW

**Comment by umanwizard** [comment]
Sounds compelling, but from a Chesterson's fence perspective, why did we ban it in the first place?

**Comment by guswynn** [comment]
@umanwizard https://github.com/MaterializeInc/materialize/pull/11439/files#diff-1040309c64844eb1b6b63d8fd67938adbf9461f1b3c61f12cf738f064a02d3deL64-L65

Basically, that at the time, some people thought that the DSL was not worth it, and manual impls were clearer

this pr posits that general personal preference (pro thiserror preference), and the size and number of error's we have outweigh those original concerns

File: src/dataflow/src/source/s3.rs
Comment (guswynn) [inline src/dataflow/src/source/s3.rs:623]: note the missing `source` impl here for the the underlying `Sdk` AND `io` errors

**Comment by benesch** [comment]
> Sounds compelling, but from a Chesterson's fence perspective, why did we ban it in the first place?

`thiserror` is a complex proc macro that, as with all proc macros, is essentially an extension to the Rust language that everyone needs to familiarize themselves with. To understand the behavior of an error type defined with `thiserror`, you need to go read the documentation on `thiserror`, instead of simply reading the implementations of the standard `From` and `fmt::Display` traits.

In particular, the formatting string syntax is _not quite_ the standard `format!` syntax, because you can use `.var`/`.0` to refer to fields of the tuple/struct to which the `error` attribute is attached. It gets increasingly hard to read the more complicated the format string gets, because you no longer to imperatively build up the format string via repeated calls to `f.write_str`: https://github.com/MaterializeInc/materialize/pull/11439/files#diff-36f5777f92061a579c5d0203815c7ab9fcc0f3996b5645e7bc0c652304d1d414R64-R72

So the tl;dr is that I find the way that `thiserror` commingles the presentation and construction of the error with the definition of the error type to make code less clear, not more clear. 

All that said, @guswynn, can you signal boost this in #eng-announce and put this to a vote? If I'm in a serious minority, I'm happy to be overruled. CC @frankmcsherry too, though, who generally distrusts proc macros.

**Comment by aalexandrov** [comment]
I don't have strong feelings about this PR, although in general I like the fact it reduced a lot of boilerplate.

However, I want to give a heads up that I am currently working my way through `scalar::EvalError` as I'm trying to add protobuf support for that type (through manually added `prost` annotations). I would therefore appreciate if we can exclude this type from any `thiserror`-related refactoring until I am done with it. The PR for that file should be out today.

**Comment by petrosagg** [comment]
I like `thiserror` a lot, especially for the simple error types where there are just some variants, potentially holding an other error in them and some message. Maybe there is an argument to be made for super complex error types in our codebase but that's usually the exception. 

The reduce of boilerplate demonstrated in this diff is also nice, I found myself doing repetitive typing recently when I had to write both the doc comments for the error variant and practically identical text for the display. We also have a lot of `impl Error for PlanError {}` blocks in our codebase that simply skip wiring up `fn source(..)`.

I would definitely vote for an unbanning of the crate and maybe for the error types that are very complex we write it manually, which is a similar approach to writing a manual `Debug` implementation when the derived one doesn't work that well.

**Comment by frankmcsherry** [comment]
I loathe Rust macros, and the "boilerplate" that is removed is invariably the "clear communication about what the code does" such that when it doesn't work as expected, folks who are not familiar with the particular macro's sublanguage are boned. Our engineering goal here is not to write the shortest code possible, but the code that retains the most value when transferred to other people.

**Comment by guswynn** [comment]
Two more positives:
- consistent with other rust repos under materialize/ (like the k8s-eip-operator)
- we wont require the use of thiserror for error types, just allow it

**Comment by guswynn** [comment]
@benesch I pushed a commit that shows that for complex variants, like `AmbiguousRename`, you can break stuff out into their own impls if need be

**Review by benesch — COMMENTED** [review]
It seems like there's a pretty clear majority in favor of allowing `thiserror`, so I'm in favor of moving forward with this PR modulo comments. Moving forward depends on the strength of @frankmcsherry's opposition, though!

**Comment by guswynn** [comment]
@frankmcsherry this pr should be up-to-date with @benesch's comments, waiting for your sign-off!

File: src/coord/src/catalog/error.rs
Comment (benesch) [inline src/coord/src/catalog/error.rs:94]:
```suggestion
```

**Review by benesch — APPROVED** [review]
Just chatted with Frank and we're good to go here, provided the `thiserror` champions (you, @alex-hunt-materialize, @petrosagg, possibly others) are willing to be on the hook for helping other folks understood any `thiserror`-related issues. Thanks for the work here, @guswynn.

**Comment by petrosagg** [comment]
That sounds great, happy to help with any issues

**Pull request merged**
